File: src/lightning.rs
use std::{collections::HashMap, str::FromStr};

use anyhow::Result;
use lightning_invoice::Invoice;
use serde::{Deserialize, Serialize};

use crate::{
    constants::LNDHUB_ENDPOINT,
    util::{get, post_json_auth},
};

/// Lightning wallet credentials
#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct Credentials {
    pub username: String,
    pub password: String,
}

/// Wallet creation response
#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(untagged)]
pub enum CreateWalletResponse {
    Username { username: String },
    Error { error: String },
}

/// Auth response
#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(untagged)]
pub enum AuthResponse {
    Result { refresh: String, token: String },
    Error { error: String },
}

/// Amount and currency
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct Money {
    pub value: String,
    pub currency: String,
}

/// Add Invoice response
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AddInvoiceResponse {
    pub req_id: String,
    pub uid: u32,
    pub payment_request: Option<String>,
    pub meta: Option<String>,
    pub metadata: Option<String>,
    pub amount: Money,
    pub rate: Option<String>,
    pub currency: String,
    pub target_account_currency: Option<String>,
    pub account_id: Option<String>,
    pub error: Option<String>,
    pub fees: Option<String>,
}

/// User balance response
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BalancesResponse {
    pub uid: u32,
    pub accounts: HashMap<String, Account>,
    pub error: Option<String>,
}

/// User account
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Account {
    pub account_id: String,
    pub balance: String,
    pub currency: String,
}

/// Pay Invoice request
#[derive(Debug, Serialize, Deserialize)]
pub struct PayInvoiceRequest {
    pub payment_request: String,
}

/// Lightning transaction
#[derive(Debug, Serialize, Deserialize)]
pub struct Transaction {
    pub txid: String,
    pub fee_txid: Option<String>,
    pub outbound_txid: Option<String>,
    pub inbound_txid: Option<String>,
    pub created_at: u64,
    pub outbound_amount: String,
    pub inbound_amount: String,
    pub outbound_account_id: String,
    pub inbound_account_id: String,
    pub outbound_uid: u32,
    pub inbound_uid: u32,
    pub outbound_currency: String,
    pub inbound_currency: String,
    pub exchange_rate: String,
    pub tx_type: String,
    pub fees: String,
    pub reference: Option<String>,
}

/// Pay invoice response
#[derive(Debug, Serialize, Deserialize, PartialEq, Eq)]
pub struct PayInvoiceResponse {
    pub payment_hash: String,
    pub uid: u32,
    pub success: bool,
    pub currency: String,
    pub payment_request: Option<String>,
    pub amount: Option<Money>,
    pub fees: Option<Money>,
    pub error: Option<String>,
    pub payment_preimage: Option<String>,
    pub destination: Option<String>,
    pub description: Option<String>,
}

/// Check payment response
#[derive(Debug, Serialize, Deserialize)]
pub struct CheckPaymentResponse {
    paid: bool,
}

/// Swap BTC onchain to Lightning response
#[derive(Debug, Serialize, Deserialize)]
pub struct SwapBtcLnResponse {
    pub address: String,
    pub commitment: String,
    pub signature: String,
    pub secret_access_key: String,
}

/// Speed of the onchain transaction
#[derive(Serialize, Deserialize, Debug)]
pub enum OnchainSpeed {
    Fast,
    Medium,
    Slow,
}

/// Swap Lightning to BTC onchain request
#[derive(Serialize, Deserialize, Debug)]
pub struct SwapLnBTCRequest {
    pub amount: u64,
    pub address: String,
    pub speed: Option<OnchainSpeed>,
}

/// Swap Lightning to BTC onchain response
#[derive(Debug, Serialize, Deserialize)]
pub struct SwapLnBtcResponse {
    pub bolt11_invoice: String,
    pub fee_sats: u32,
}

/// Creates a new lightning custodial wallet
pub async fn create_wallet(username: &str, password: &str) -> Result<CreateWalletResponse> {
    let endpoint = LNDHUB_ENDPOINT.read().await;
    let creds = Credentials {
        username: username.to_string(),
        password: password.to_string(),
    };
    let create_url = format!("{endpoint}/create");
    let response = post_json_auth(&create_url, &Some(creds), None).await?;

    let res: CreateWalletResponse = serde_json::from_str(&response)?;

    Ok(res)
}

/// Get a auth tokens
pub async fn auth(username: &str, password: &str) -> Result<AuthResponse> {
    let creds = Credentials {
        username: username.to_string(),
        password: password.to_string(),
    };
    let endpoint = LNDHUB_ENDPOINT.read().await;
    let auth_url = format!("{endpoint}/auth");
    let response = post_json_auth(&auth_url, &Some(creds), None).await?;
    let response: AuthResponse = serde_json::from_str(&response)?;

    Ok(response)
}

/// Creates a lightning invoice
pub async fn create_invoice(
    description: &str,
    amount: u32,
    token: &str,
) -> Result<AddInvoiceResponse> {
    let endpoint = LNDHUB_ENDPOINT.read().await;
    let amount = amount as f32 / 100_000_000.0;
    let amt_str = amount.to_string();
    let url = format!("{endpoint}/addinvoice?amount={amt_str}&meta={description}");
    let response = get(&url, Some(token)).await?;
    let invoice: AddInvoiceResponse = serde_json::from_str(&response)?;

    Ok(invoice)
}

/// Decode a lightning invoice (bolt11)
pub fn decode_invoice(payment_request: &str) -> Result<Invoice> {
    let invoice = Invoice::from_str(payment_request)?;

    Ok(invoice)
}

/// Get user lightning balance
pub async fn get_balance(token: &str) -> Result<Vec<Account>> {
    let endpoint = LNDHUB_ENDPOINT.read().await;
    let url = format!("{endpoint}/balance");
    let response = get(&url, Some(token)).await?;
    let balance: BalancesResponse = serde_json::from_str(&response)?;
    let mut accounts = Vec::new();
    for (_, value) in balance.accounts {
        accounts.push(value);
    }

    Ok(accounts)
}

/// Pay a lightning invoice
pub async fn pay_invoice(payment_request: &str, token: &str) -> Result<PayInvoiceResponse> {
    let endpoint = LNDHUB_ENDPOINT.read().await;
    let url = format!("{endpoint}/payinvoice");
    let req = PayInvoiceRequest {
        payment_request: payment_request.to_string(),
    };
    let response = post_json_auth(&url, &Some(req), Some(token)).await?;
    let response: PayInvoiceResponse = serde_json::from_str(&response)?;

    Ok(response)
}

/// Get successful lightning transactions user made. Order newest to oldest.
pub async fn get_txs(token: &str) -> Result<Vec<Transaction>> {
    let endpoint = LNDHUB_ENDPOINT.read().await;
    let url = format!("{endpoint}/gettxs");
    let response = get(&url, Some(token)).await?;
    let txs = serde_json::from_str(&response)?;

    Ok(txs)
}

/// Check if a lightning invoice has been paid
pub async fn check_payment(payment_hash: &str) -> Result<bool> {
    let endpoint = LNDHUB_ENDPOINT.read().await;
    let url = format!("{endpoint}/checkpayment?payment_hash={payment_hash}");
    let response = get(&url, None).await?;
    let r = serde_json::from_str::<CheckPaymentResponse>(&response)?;

    Ok(r.paid)
}

/// Swap BTC onchain to Lightning
pub async fn swap_btc_ln(token: &str, ln_address: Option<String>) -> Result<SwapBtcLnResponse> {
    let ln_address_query = match ln_address {
        Some(a) => format!("?lnurl_or_lnaddress={}", a),
        None => "".to_string(),
    };
    let endpoint = LNDHUB_ENDPOINT.read().await;
    let url = format!("{endpoint}/get_onchain_address{ln_address_query}");
    let response = get(&url, Some(token)).await?;
    let r = serde_json::from_str::<SwapBtcLnResponse>(&response)?;

    Ok(r)
}

/// Swap Lightning to BTC onchain
pub async fn swap_ln_btc(address: &str, amount: u64, token: &str) -> Result<SwapLnBtcResponse> {
    let endpoint = LNDHUB_ENDPOINT.read().await;
    let url = format!("{endpoint}/make_onchain_swap");
    let req = SwapLnBTCRequest {
        address: address.to_string(),
        amount,
        speed: Some(OnchainSpeed::Fast),
    };
    let response = post_json_auth(&url, &Some(req), Some(token)).await?;
    let r = serde_json::from_str::<SwapLnBtcResponse>(&response)?;

    Ok(r)
}

File: lib/web/lightning.ts
// Methods meant to work with LNDHubX defined within the web::lightning module from bitmask-core:
// https://github.com/diba-io/bitmask-core/blob/development/src/web.rs

import * as BMC from "./bitmask_core";

export const createWallet = async (
  username: string,
  password: string
): Promise<CreateWalletResponse> =>
  JSON.parse(await BMC.create_wallet(username, password));

export const auth = async (
  username: string,
  password: string
): Promise<AuthResponse> => JSON.parse(await BMC.auth(username, password));

export const createInvoice = async (
  description: string,
  amount: number,
  token: string
): Promise<AddInvoiceResponse> =>
  JSON.parse(await BMC.ln_create_invoice(description, amount, token));

export const getBalance = async (token: string): Promise<Account> =>
  JSON.parse(await BMC.get_balance(token));

export const getTxs = async (token: string): Promise<LnTransaction[]> =>
  JSON.parse(await BMC.get_txs(token));

export const payInvoice = async (
  paymentRequest: string,
  token: string
): Promise<PayInvoiceResponse> =>
  JSON.parse(await BMC.pay_invoice(paymentRequest, token));

export const checkPayment = async (
  paymentHash: string
): Promise<CheckPaymentResponse> =>
  JSON.parse(await BMC.check_payment(paymentHash));

export const swapBtcLn = async (token: string): Promise<SwapBtcLnResponse> =>
  JSON.parse(await BMC.swap_btc_ln(token));

export const swapLnBtc = async (
  address: string,
  amount: bigint,
  token: string
): Promise<SwapLnBtcResponse> =>
  JSON.parse(await BMC.swap_ln_btc(address, amount, token));

// Core type interfaces based on structs defined within the bitmask-core Rust crate:
// https://github.com/diba-io/bitmask-core/blob/development/src/structs.rs

export interface LnCredentials {
  login: string;
  password: string;
  refreshToken: string;
  accessToken: string;
}

// Lndhubx Create wallet endpoint response
export interface CreateWalletResponse {
  username?: string;
  error?: string;
}

// lndhubx Auth response
export type AuthResponse = ErrorResponse | AuthResponseOk;

export interface AuthResponseOk {
  refresh: string;
  token: string;
}

export interface ErrorResponse {
  error: string;
}

// User Account
export interface Account {
  account_id: string;
  balance: string;
  currency: string;
}

// Amount and currency
export interface Money {
  value: string;
  currency: string;
}

// Lndhubx Add invoice endpoint response
export interface AddInvoiceResponse {
  req_id: string;
  uid: number;
  payment_request: string;
  meta: string;
  metadata: string;
  amount: Money;
  rate: string;
  currency: string;
  target_account_currency: string;
  account_id: string;
  error: string;
  fees: string;
}

// Lndhubx lightning transaction
export interface LnTransaction {
  txid: string;
  fee_txid: string;
  outbound_txid: string;
  inbound_txid: string;
  created_at: bigint;
  date: number;
  outbound_amount: string;
  inbound_amount: string;
  outbound_account_id: string;
  inbound_account_id: string;
  outbound_uid: number;
  inbound_uid: number;
  outbound_currency: string;
  inbound_currency: string;
  exchange_rate: string;
  tx_type: string;
  fees: string;
  reference: string;
}

export interface LnWalletData {
  balance: Account;
  transactions: LnTransaction[];
}

// Lndhubx Pay invoice response
export interface PayInvoiceResponse {
  payment_hash: string;
  uid: number;
  success: boolean;
  currency: string;
  payment_request: string;
  amount: Money;
  fees: Money;
  error: string;
  payment_preimage: string;
  destination: string;
  description: string;
}

// Lndhubx Check payment response
export interface CheckPaymentResponse {
  paid: boolean;
}

export interface SwapBtcLnResponse {
  address: string;
  commitment: string;
  signature: string;
  secret_access_key: string;
}

export interface SwapLnBtcResponse {
  bolt11_invoice: string;
  fee_sats: number;
}
